tidy3d/components/parameter_perturbation.py

```diff
@@ -0,0 +1,722 @@
+        if self.temperature_range is None:
+            return sample(self, temperature)
+
+        temh_min, temh_max = self.temperature_range
+        if np.any(temperature < temh_min) or np.any(temperature > temh_max):
+            log.warning(
+                "temperature passed to 'HeatPerturbation.sample()'"
+                f"is outside of 'HeatPerturbation.temperature_range' = {self.temperature_range}"
+            )
+        return sample(self, temperature)
+
+    return _sample
+
+
+class HeatPerturbation(AbstractPerturbation):
+    """Abstract class for heat perturbation."""
+
+    temperature_range: Tuple[pd.NonNegativeFloat, pd.NonNegativeFloat] = pd.Field(
+        (0, inf),
+        title="Temperature range",
+        description="Temparature range in which perturbation model is valid.",
+        units=KELVIN,
+    )
+
+    @abstractmethod
+    def sample(
+        self, temperature: Union[ArrayLike[float], SpatialDataArray]
+    ) -> Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]:
+        """Sample perturbation.
+
+        Parameters
+        ----------
+        temperature : Union[ArrayLike[float], SpatialDataArray]
+            Temperature sample point(s).
+
+        Returns
+        -------
+        Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]
+            Sampled perturbation value(s).
+        """
+
+    @add_ax_if_none
+    def plot(
+        self,
+        temperature: ArrayLike[float],
+        val: FieldVal = "real",
+        ax: Ax = None,
+    ) -> Ax:
+        """Plot perturbation using provided temperature sample points.
+
+        Parameters
+        ----------
+        temperature : ArrayLike[float]
+            Array of temperature sample points.
+        val : Literal['real', 'imag', 'abs', 'abs^2', 'phase'] = 'real'
+            Which part of the field to plot.
+        ax : matplotlib.axes._subplots.Axes = None
+            Matplotlib axes to plot on, if not specified, one is created.
+
+        Returns
+        -------
+        matplotlib.axes._subplots.Axes
+            The supplied or created matplotlib axes.
+        """
+
+        temperature_numpy = np.array(temperature)
+
+        values = self.sample(temperature_numpy)
+        values = self._get_val(values, val)
+
+        ax.plot(temperature_numpy, values)
+        ax.set_xlabel("temperature (K)")
+        ax.set_ylabel(f"{val}(perturbation value)")
+        ax.set_title("temperature dependence")
+        ax.set_aspect("auto")
+
+        return ax
+
+
+class LinearHeatPerturbation(HeatPerturbation):
+    """Linear heat perturbation."""
+
+    temperature_ref: pd.NonNegativeFloat = pd.Field(
+        ...,
+        title="Reference temperature",
+        description="Reference temperature.",
+        units=KELVIN,
+    )
+
+    coeff: Union[float, Complex] = pd.Field(
+        ...,
+        title="Thermo-optic Coefficient",
+        description="Thermo-optic Coefficient.",
+    )
```

Comment:
units? seems a lot of the other ones might be missing them as well

Reply:
the units are actually unknown here because it can be applied to any parameter, but I set it to `1/K` as in case of a dimensionless parameter